File: bindings/php/examples/batch_lookup.php
<?php
/**
 * batch_lookup.php - Batch lookup operations with liblpm
 *
 * This example demonstrates efficient batch lookups for high-throughput
 * scenarios where you need to look up many addresses at once.
 *
 * Usage:
 *   php -d extension=liblpm.so batch_lookup.php
 */

echo "=== liblpm PHP Extension - Batch Lookup Example ===\n\n";

if (!extension_loaded('liblpm')) {
    die("Error: liblpm extension is not loaded.\n");
}

// ============================================================================
// Create and populate table
// ============================================================================
echo "--- Setting up routing table ---\n";
$table = new LpmTableIPv4();

// Insert a variety of routes
$routes = [
    "0.0.0.0/0" => 1,       // Default
    "10.0.0.0/8" => 100,
    "10.1.0.0/16" => 110,
    "10.2.0.0/16" => 120,
    "172.16.0.0/12" => 200,
    "192.168.0.0/16" => 300,
    "192.168.1.0/24" => 310,
    "192.168.2.0/24" => 320,
];

foreach ($routes as $prefix => $nextHop) {
    $table->insert($prefix, $nextHop);
}
echo "Inserted " . count($routes) . " routes\n\n";

// ============================================================================
// Prepare batch of addresses
// ============================================================================
echo "--- Preparing batch of addresses ---\n";

$addresses = [
    // 10.x.x.x range
    "10.0.0.1",
    "10.1.1.1",
    "10.2.2.2",
    "10.3.3.3",
    
    // 172.x.x.x range
    "172.16.1.1",
    "172.20.5.10",
    "172.31.255.255",
    
    // 192.168.x.x range
    "192.168.0.1",
    "192.168.1.100",
    "192.168.2.50",
    "192.168.3.1",
    
    // Public addresses (default route)
    "8.8.8.8",
    "1.1.1.1",
    "208.67.222.222",
];

echo "Batch size: " . count($addresses) . " addresses\n\n";

// ============================================================================
// Perform batch lookup
// ============================================================================
echo "--- Batch lookup ---\n";

// Time the batch lookup
$startTime = microtime(true);
$results = $table->lookupBatch($addresses);
$endTime = microtime(true);

// Display results
foreach ($addresses as $i => $addr) {
    $nh = $results[$i];
    if ($nh === false) {
        printf("  %-20s -> NO MATCH\n", $addr);
    } else {
        printf("  %-20s -> %d\n", $addr, $nh);
    }
}

$elapsed = ($endTime - $startTime) * 1000000; // microseconds
printf("\nBatch lookup completed in %.2f microseconds\n", $elapsed);
printf("Average: %.2f microseconds per lookup\n\n", $elapsed / count($addresses));

// ============================================================================
// Compare with individual lookups
// ============================================================================
echo "--- Comparison: Individual lookups ---\n";

$startTime = microtime(true);
$individualResults = [];
foreach ($addresses as $addr) {
    $individualResults[] = $table->lookup($addr);
}
$endTime = microtime(true);

$elapsed = ($endTime - $startTime) * 1000000;
printf("Individual lookups completed in %.2f microseconds\n", $elapsed);
printf("Average: %.2f microseconds per lookup\n\n", $elapsed / count($addresses));

// Verify results match
$match = $results === $individualResults;
echo "Results match: " . ($match ? "YES" : "NO") . "\n\n";

// ============================================================================
// Large batch example
// ============================================================================
echo "--- Large batch example ---\n";

// Generate many random addresses
$largeAddresses = [];
for ($i = 0; $i < 1000; $i++) {
    $largeAddresses[] = sprintf("%d.%d.%d.%d",
        rand(0, 255), rand(0, 255), rand(0, 255), rand(0, 255));
}

echo "Generated " . count($largeAddresses) . " random addresses\n";

$startTime = microtime(true);
$largeResults = $table->lookupBatch($largeAddresses);
$endTime = microtime(true);

$elapsed = ($endTime - $startTime) * 1000000;
printf("Large batch completed in %.2f microseconds\n", $elapsed);
printf("Average: %.3f microseconds per lookup\n", $elapsed / count($largeAddresses));

// Count results
$found = count(array_filter($largeResults, fn($r) => $r !== false));
$notFound = count($largeResults) - $found;
echo "Results: $found found, $notFound not found\n\n";

// ============================================================================
// Clean up
// ============================================================================
$table->close();
echo "=== Example Complete ===\n";

File: bindings/python/examples/batch_lookup.py
#!/usr/bin/env python3
"""Example demonstrating batch lookup operations.

Batch lookups are more efficient when looking up many addresses,
as they reduce the overhead of crossing the Python/C boundary.
"""

import time
from ipaddress import IPv4Address, IPv4Network

from liblpm import LpmTableIPv4


def generate_test_routes(count: int = 1000):
    """Generate test routes for benchmarking."""
    routes = []
    for i in range(min(count, 256)):
        for j in range(min(count // 256 + 1, 256)):
            prefix = f'{i}.{j}.0.0/16'
            next_hop = i * 256 + j
            routes.append((prefix, next_hop))
            if len(routes) >= count:
                return routes
    return routes


def generate_test_addresses(count: int = 10000):
    """Generate test addresses for lookups."""
    addresses = []
    for i in range(256):
        for j in range(256):
            for k in range(count // 65536 + 1):
                addr = IPv4Address(f'{i}.{j}.{k}.1')
                addresses.append(addr)
                if len(addresses) >= count:
                    return addresses
    return addresses


def compare_single_vs_batch():
    """Compare single lookups vs batch lookups."""
    print("Comparing single lookups vs batch lookups")
    print("-" * 50)
    
    # Create and populate table
    routes = generate_test_routes(1000)
    addresses = generate_test_addresses(10000)
    
    with LpmTableIPv4() as table:
        print(f"Inserting {len(routes)} routes...")
        for prefix, next_hop in routes:
            table.insert(prefix, next_hop)
        
        print(f"Looking up {len(addresses)} addresses...")
        
        # Single lookups
        start = time.perf_counter()
        single_results = []
        for addr in addresses:
            result = table.lookup(addr)
            single_results.append(result)
        single_time = time.perf_counter() - start
        
        # Batch lookup
        start = time.perf_counter()
        batch_results = table.lookup_batch(addresses)
        batch_time = time.perf_counter() - start
        
        # Verify results match
        assert len(single_results) == len(batch_results)
        mismatches = sum(1 for s, b in zip(single_results, batch_results) if s != b)
        
        print(f"\nResults:")
        print(f"  Single lookups: {single_time:.4f}s ({len(addresses)/single_time:.0f} lookups/s)")
        print(f"  Batch lookup:   {batch_time:.4f}s ({len(addresses)/batch_time:.0f} lookups/s)")
        print(f"  Speedup:        {single_time/batch_time:.2f}x")
        print(f"  Results match:  {'Yes' if mismatches == 0 else f'No ({mismatches} mismatches)'}")


def batch_with_mixed_results():
    """Demonstrate batch lookup with mixed hit/miss results."""
    print("\n" + "-" * 50)
    print("Batch lookup with mixed hit/miss results")
    print("-" * 50)
    
    with LpmTableIPv4() as table:
        # Insert a few specific routes
        table.insert('192.168.0.0/16', 100)
        table.insert('10.0.0.0/8', 200)
        
        # Mix of addresses that will/won't match
        addresses = [
            IPv4Address('192.168.1.1'),   # Hit -> 100
            IPv4Address('8.8.8.8'),        # Miss
            IPv4Address('10.0.0.1'),       # Hit -> 200
            IPv4Address('172.16.0.1'),     # Miss
            IPv4Address('192.168.2.1'),   # Hit -> 100
        ]
        
        results = table.lookup_batch(addresses)
        
        print("Batch lookup results:")
        for addr, result in zip(addresses, results):
            status = f"next_hop={result}" if result is not None else "no match"
            print(f"  {addr} -> {status}")


def main():
    """Run batch lookup examples."""
    print("liblpm Python Bindings - Batch Lookup Example")
    print("=" * 50)
    print()
    
    compare_single_vs_batch()
    batch_with_mixed_results()
    
    print("\n" + "=" * 50)
    print("Batch lookup examples completed!")


if __name__ == '__main__':
    main()
